### PR TITLE
instancetypes: Improve instancetype snapshots

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
The `ControllerRevisions` used for instancetype and preference snapshots will contain the whole object.
It is useful, because the object contains version information, and we can use methods in the `runtime` package to decode it.

**Release note**:
```release-note
Improves instancetype and preference controller revisions. This is a backwards incompatible change and introduces a new v1alpha2 api for instancetype and preferences.
```

### CODE REVIEW

File: tests/instancetype_test.go
Comment (lyarwood) [inline tests/instancetype_test.go:223]: https://github.com/kubevirt/kubevirt/pull/8280 ^_^ I honestly don't mind if you just amend that entire commit in here tbh. Not all of the tests using the disk actually required it FWIW.

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:257]: Can you drop the `and store the name below` from this comment, it doesn't make sense now given it has moved.

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:131]: I forget why I did this but it shouldn't be possible to have a plural kind here right?

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:331]:
```
case *instancetypev1alpha1.VirtualMachinePreference, *instancetypev1alpha1.VirtualMachineClusterPreference:
    return &obj.Spec, nil
```

**Review by lyarwood — CHANGES_REQUESTED** [review]
Looking good overall, really like the new ControllerRevision code so thanks for that!

I've listed a few nits and a conflict with a PR I posted earlier today. If you could also add some more context to some of your refactors I'd appreciate it. I get that it's boring but it's useful for reviewers and the original author (./me waves) to understand why you wanted to change certain things.

Cheers!

**Comment by lyarwood** [comment]
Odd this is also causing the following failure for me locally:

```
# ./cluster-up/kubectl.sh apply -f ./examples/csmall.yaml -f ./examples/vm-cirros-csmall.yaml
[..]
# ./cluster-up/virtctl.sh start vm-cirros-csmall
[..]
# ./cluster-up/kubectl.sh get vms/vm-cirros-csmall -o json | jq .status.conditions[1]
selecting docker as container runtime
{
  "lastProbeTime": null,
  "lastTransitionTime": "2022-08-11T14:50:02Z",
  "message": "Failure while starting VMI: admission webhook \"virtualmachine-validator.kubevirt.io\" denied the request: Failure to find instancetype: failed to decode object in ControllerRevision: Object 'Kind' is missing in '{\"metadata\":{\"name\":\"csmall\",\"namespace\":\"default\",\"uid\":\"f2218614-7e54-4ebe-8c7f-fec60b0d7d86\",\"resourceVersion\":\"87367\",\"generation\":1,\"creationTimestamp\":\"2022-08-11T14:49:54Z\"},\"spec\":{\"cpu\":{\"guest\":1},\"memory\":{\"guest\":\"128Mi\"}}}'",
  "reason": "FailedCreate",
  "status": "True",
  "type": "Failure"
}
```

**Comment by lyarwood** [comment]
> > Odd this is also causing the following failure for me locally:
> 
> Maybe the code is not working yet. I'm still compiling it and uploading to quay to test it. I will mark this as draft in the meantime.

Yeah it reprodiuces in the unit tests as well FWIW. I think we need something like this as TypeMeta isn't present :

https://github.com/kubernetes/kubernetes/issues/3030#issuecomment-700099699

**Comment by lyarwood** [comment]
> > > Odd this is also causing the following failure for me locally:
> > 
> > 
> > Maybe the code is not working yet. I'm still compiling it and uploading to quay to test it. I will mark this as draft in the meantime.
> 
> Yeah it reproduces in the unit tests as well FWIW. I think we need something like this as TypeMeta isn't present :
> 
> [kubernetes/kubernetes#3030 (comment)](https://github.com/kubernetes/kubernetes/issues/3030#issuecomment-700099699)

The following is working for me now:

https://gist.github.com/lyarwood/1645a40c115eb62e6da37ca1e289be9e

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:424]: Is there a reason we can't collapse these cases into one?

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [inline pkg/instancetype/instancetype.go:69]: This really doesn't strike me as great practice even if as you say we can only get here through some dev failure at the moment. Do we really want to panic at all in `virt-controller` or `virt-api` ?

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:139]: Again, might be something to cleanup outside of this series but I don't think we can end up with a plural kind here.

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:264]: Again, please amend the comment given the changes in the series.

File: tests/instancetype_test.go
Comment (lyarwood) [inline tests/instancetype_test.go:567]: We also spoke about adding functional test coverage of the duplicate CR detection code in `storeRevision` [1] given the issues we saw with the previous PR revision and differences testing it with fakeclients, can you add something here?

[1] https://github.com/kubevirt/kubevirt/blob/ebeed7540784cbdd05b11df99e1fc2cd7c5597fc/pkg/instancetype/instancetype.go#L254-L268

**Review by lyarwood — CHANGES_REQUESTED** [review]
Looking good but I've still got a few nits I'd like discussed and/or addressed. Again I'd also appreciate more context in the actual commit messages justifying some of the refactors.

File: tests/instancetype_test.go
Comment (akrejcir) [reply tests/instancetype_test.go:224]: Ok, I will do it.

File: pkg/instancetype/instancetype.go
Comment (akrejcir) [reply pkg/instancetype/instancetype.go:69]: @davidvossel do you know if it is ok to use `panic` here?

File: pkg/instancetype/instancetype.go
Comment (akrejcir) [reply pkg/instancetype/instancetype.go:424]: Yes, it is not possible in golang type system. Which type would the `obj` variable have in the combined branch?
It would need to be some common interface, but then we could not access `.Spec` field.

File: pkg/instancetype/instancetype.go
Comment (akrejcir) [reply pkg/instancetype/instancetype.go:139]: Most of other kubernetes tools also accept plural when expecting Kind. So I thought it would be good to also accept it here, so user is not surprised.

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:424]: Ah apologies I see `obj` stays as `runtime.Object` without access to `.Spec`, I just recalled that you can match on multiple types and didn't think about access to `.Spec` via `obj` here.

File: pkg/instancetype/instancetype.go
Comment (lyarwood) [reply pkg/instancetype/instancetype.go:139]: ACK.

File: pkg/instancetype/instancetype.go
Comment (davidvossel) [reply pkg/instancetype/instancetype.go:69]: in general, i typically see `panic()` used during initialization, when some initialization step fails and there's no alternative but to exit. So the panic is useful because it prevents either unrecoverable or undefined behavior.

I also occasionally see functions called `<DoSomething>OrDie()`, like fields.ParseSelectorOrDie() for example. That `OrDie` part is kind of signifying to the caller that this is either going to pass, or kill your process. I wouldn't say that's super common.

Since the functionality in question is occurring after initialization within the reconcile loop or during a subresource http request, I'd be inclined to handle it as an error rather than panic.

Comment (akrejcir) [reply pkg/instancetype/instancetype.go:69]: Thank you for the explanation, I will change it to error return.

File: tests/instancetype_test.go
Comment (akrejcir) [reply tests/instancetype_test.go:567]: Done

File: pkg/instancetype/instancetype.go
Comment (akrejcir) [reply pkg/instancetype/instancetype.go]: Done

**Review by lyarwood — APPROVED** [review]
/lgtm

Thanks for the changes, this LGTM now. The functional test isn't what I expected but you're just allowing the controller to create the initial CR before using that as the unexpected CR once the VM is recreated so it still exercises the codepath.

**Comment by lyarwood** [comment]
/lgtm cancel

Apologies I forgot you can't run tests locally at the moment, there are still some issues to work out with the functional tests.

**Comment by lyarwood** [comment]
/lgtm 

This should be passing now that we have resolved the issue with the functional tests.

**Comment by akrejcir** [comment]
@davidvossel or @vladikr , can you take a look when you a have a moment, and approve, if it is ok?

File: staging/src/kubevirt.io/api/instancetype/v1alpha1/types.go
Comment (lyarwood) [inline staging/src/kubevirt.io/api/instancetype/v1alpha1/types.go:77]: @davidvossel Moving the discussion here, this will break existing users but as we are still in alpha breaking changes like this are still accepted AFAIK [1]. We could signal this with a bump to `v1alpha2` if it would help?

[1] https://kubernetes.io/docs/reference/using-api/#api-versioning

Comment (lyarwood) [reply staging/src/kubevirt.io/api/instancetype/v1alpha1/types.go:77]: FWIW I've also created https://github.com/kubevirt/kubevirt/issues/8235 recently to track getting this API to beta.

Comment (akrejcir) [reply staging/src/kubevirt.io/api/instancetype/v1alpha1/types.go:77]: Increasing the version to `v1alpha2` would not solve the problem that existing ControllerRevisions would not work.
But because this is still alpha feature, could we document somewhere that after this change, old ControllerRevisions need to be removed? Then they will be recreated in the new format.

Comment (lyarwood) [reply staging/src/kubevirt.io/api/instancetype/v1alpha1/types.go:77]: The bump is just to signal that something had changed but yeah I agree that a release note would be useful here.

**Comment by davidvossel** [comment]
> The ControllerRevisions used for instancetype and preference snapshots will contain the whole object.
It is useful, because the object contains version information, and we can use methods in the runtime package to decode it.

@akrejcir do we not already have the version information in the snapshot today? Would you be able to expand a bit on how this change improves the situation?

**Comment by akrejcir** [comment]
> @akrejcir do we not already have the version information in the snapshot today? Would you be able to expand a bit on how this change improves the situation?

Yes, there is already version information. But I think by storing the whole object in ControllerRevision we can use library methods from `k8s.io/apimachinery` to decode and convert it to newer API. 

Other minor benefits are:
- Removed API objects: `VirtualMachineInstancetypeSpecRevision`, `VirtualMachinePreferenceSpecRevision`
- When viewing the ControllerRevision using `oc get controllerrevision -o yaml`, the output is a readable indented yaml of the object it contains. The current implementation uses `[]byte` to store json encoded object, which will be displayed in base64.

**Comment by davidvossel** [comment]
> When viewing the ControllerRevision using oc get controllerrevision -o yaml, the output is a readable indented yaml of the object it contains. The current implementation uses []byte to store json encoded object, which will be displayed in base64.

oh, that's kind of neat to know, thanks!

File: staging/src/kubevirt.io/api/instancetype/v1alpha1/types.go
Comment (davidvossel) [reply staging/src/kubevirt.io/api/instancetype/v1alpha1/types.go:77]: yeah, if we're goign to do this, we need to bump the api version v1alpha2 and have a release note. That should be fine.

**Review by davidvossel — COMMENTED** [review]
can we bump the api to v1alpha2 in order to indicate this is a backwards incompatible change?

**Comment by akrejcir** [comment]
Rebased and bumped the API version.

**Review by davidvossel — COMMENTED** [review]
looks good, but shouldn't we remove the v1alpha1 api now too?

**Review by davidvossel — COMMENTED** [review]
/approve
/lgtm

nevermind about removing v1alpha1 for now. I see you created v1alpha2 as the storage version.I think it's okay to remove v1alpha1 after the next kubevirt release

**Review by davidvossel — COMMENTED** [review]
/hold
/lgtm cancel

I think we still have an issue with versioning here. If someone updates from using v1alpha1 of instancetypes, then the VM reconciler will fail because the unmarshalling/decoding of the existing controller revision will fail.

It's okay if we break instancetypes with a new v1alpha2. But as this is now, I'm not certain that there's a way to use VMs with existing v1alpha1 controller revisions after a kubevirt update... Basically if you have a VM before the kubevirt update, it will have a controller revsion. After the update that controller revision will get retrieved by the vm controller, but cause a reconcile failure. This will cause the VM to never be able to be started, stopped, etc...

**Comment by lyarwood** [comment]
> It's okay if we break instancetypes with a new v1alpha2. But as this is now, I'm not certain that there's a way to use VMs with existing v1alpha1 controller revisions after a kubevirt update... Basically if you have a VM before the kubevirt update, it will have a controller revsion. After the update that controller revision will get retrieved by the vm controller, but cause a reconcile failure. This will cause the VM to never be able to be started, stopped, etc...

So this really depends on our interpretation of the `backwards incompatible` term for `alpha` versions. While what you say is true if we document how a user can correct this by deleting the underlying CR and removing the reference from the VM I think that the behaviour is acceptable for this version. I don't think it would be acceptable for any `beta` or `ga` version to not automatically handle previous versions however but for `alpha` I think we are good.

> **Release note**:
> 
> ```
> Improves instancetype and preference controller revisions. This is a backwards incompatible change and introduces a new v1alpha2 api for instancetype and preferences.
> ```

As above can we include instructions for existing VMs and how to correct the older CRs?

**Comment by akrejcir** [comment]
Added code to read `ControllerRevisions` containing `v1alpha1` instancetype data and convert it to `v1alpha2` format.

**Comment by lyarwood** [comment]
> I guess we can also mark the `v1alpah1` CRD version as deprecated now as well?

Ignore this for now, we can just mark both as deprecated once we have `v1beta`. This LGTM but I'm not going to vote as the `approved` label is still present. @davidvossel I'll leave this for you to approve when you have time to review.

**Comment by lyarwood** [comment]
/lgtm
/hold

@davidvossel this LGTM now, do you have any other comments or concerns?

**Review by davidvossel — COMMENTED** [review]
/hold cancel

thanks for the compatibility conversion!

**Comment by lyarwood** [comment]
/lgtm

Fun times, c9e474a6e9be9daa8f9a35e191d00508880f0b9a landed overnight and added more tests using the `v1alpha1` version causing various failures in the gate. I've rebased the series on main again and moved all remaining tests to `v1alpha2` in 1149942609623dffc30cbadca0b32679bd7e8134. This should be good now so I've added `lgtm` back in.

**Comment by lyarwood** [comment]
/lgtm

Apologies for the noise I missed the [BUILD.bazel](https://github.com/kubevirt/kubevirt/commit/2f4fda4faf62279112e8c28d6d7df01633011ee9#diff-d95f450f91602959c2333917cb40190bceb34011351601f6fe8bda1991534c92) changes, should be good now.

**Pull request merged**
